### ISSUE TITLE
[Android] 'When' in a catch statement, crashes app

### ISSUE THREAD

**Issue by thudugala** [xamarin/Xamarin.Forms#6269]
### Description

If exception is catch with 'When' contextual keyword, the application crashes. 

```cs
try
{
    throw new Exception("Test");
}
catch (Exception ex) when (ex.Message == "Test")
{
    Console.WriteLine("Exception Test");
}
```

### Steps to Reproduce

1.  Run the attached project 
2.  Tap, 'Try Exception' button

### Expected Behavior

app should **not** crash. 

### Actual Behavior

App crashes. 

##### Native Crash Reporting

Got a SIGSEGV while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.

### Basic Information

- Version with issue: Visual Studio 16.1
- Last known good version: Visual Studio 16.0
- Platform Target Frameworks: 
  - Android: Andoid 9.0
- Nuget Packages: Xmarin.Forms 3.6.0.344457

### Reproduction Link

[App2.zip](https://github.com/xamarin/Xamarin.Forms/files/3209341/App2.zip)

**Comment by amirvenus** [comment]
Have you tried changing 

`(ex.Message == "Test")`

to 

`(ex?.Message == "Test")`

**Comment by thudugala** [comment]
@amirvenus `ex` will never be null.

![image](https://user-images.githubusercontent.com/4112014/58283800-61a65780-7dfd-11e9-978e-95d1805eaac3.png)

**Comment by brendanzagaeski** [comment]
Thanks for the report and the test case! I can confirm that this behavior around exception filters also has the same native backtrace as <https://github.com/xamarin/xamarin-android/issues/3112> and <https://github.com/mono/mono/issues/14170>, so I will close this Xamarin.Forms issue as a duplicate. The fix will require a new version of Xamarin.Android that will be based on a new version of the Mono runtime and debugger. (The Xamarin.Forms NuGet package version will not affect this behavior.)

For reference, here's the top of the native backtrace with symbols that I got from the test case in the [description](https://github.com/xamarin/Xamarin.Forms/issues/6269#issue-447345745):

```
  * frame #0: 0x0000007126cf5da8 libart.so`art_sigsegv_fault
    frame #1: 0x0000007126cf62b4 libart.so`art::FaultManager::HandleFault(int, siginfo*, void*) + 356
    frame #2: 0x00000058b5382b5c app_process64`___lldb_unnamed_symbol20$$app_process64 + 572
    frame #3: 0x00000071acbdd870 [vdso]
    frame #4: 0x00000071100cd640 libmonosgen-2.0.d.so`mono_handle_exception_internal [inlined] jinfo_get_method(ji=0x0000000000000000) at mini-runtime.h:390
    frame #5: 0x00000071100cd638 libmonosgen-2.0.d.so`mono_handle_exception_internal(ctx=<unavailable>, obj=<unavailable>, resume=0, out_ji=<unavailable>) at mini-exceptions.c:2701
    frame #6: 0x00000071100cd640 libmonosgen-2.0.d.so`mono_handle_exception_internal [inlined] jinfo_get_method(ji=0x000000710a02faa8) at mini-runtime.h:390
    frame #7: 0x00000071100cd638 libmonosgen-2.0.d.so`mono_handle_exception_internal(ctx=<unavailable>, obj=<unavailable>, resume=0, out_ji=<unavailable>) at mini-exceptions.c:2701
    frame #8: 0x00000071100cd2b8 libmonosgen-2.0.d.so`mono_handle_exception(ctx=<unavailable>, obj=<unavailable>) at mini-exceptions.c:2988
    frame #9: 0x000000711010501c libmonosgen-2.0.d.so`mono_arm_throw_exception(arg=<unavailable>, pc=<unavailable>, int_regs=0x0000007fcd282ab0, fp_regs=0x0000007fcd282bb0, corlib=<unavailable>, rethrow=0) at exceptions-arm64.c:410
```

**Comment by thudugala** [comment]
Issue still exist in VS 16.1.1

**Comment by tobiasschulz** [comment]
How is this not fixed yet? It is impossible to build anything using the STABLE version and a downgrade is NOT possible with the community version. Seriously?

**Comment by thudugala** [comment]
fixed in 16.1.2

**Comment by tipa** [comment]
still crashes for me...
